### PR TITLE
Make sure not to fail in for loop sending to takers

### PR DESCRIPTION
**Pull request by da-kami**

### CODE REVIEW

File: daemon/src/maker_inc_connections.rs [hunk 1]
```diff
@@ -157,9 +157,16 @@ impl Actor {
...
-                .await?;
+            if conn
+                .do_send_async(wire::MakerToTaker::CurrentOrder(order.clone()))
+                .await
+                .is_err()
+            {
+                tracing::trace!(%taker_id, "removing outdated connection to taker because unable to send order: {:?}", order);
+                self.write_connections.remove(&taker_id);
+            } else {
+                tracing::trace!(%taker_id, "sending new order: {:?}", order);
```
Comment (bonomat) [inline daemon/src/maker_inc_connections.rs]: Nit: it happened already 😬

```suggestion
                tracing::trace!(%taker_id, "sent new order: {:?}", order);
```

**Review by bonomat — APPROVED** [review]
Did you try if this was the source of our problems?

**Comment by da-kami** [comment]
> Did you try if this was the source of our problems?

Was able to reproduce the problem locally on `master` - steps:

1. start maker
2. start taker
3. check that orders arrive on taker
4. keep maker running, restart taker
5. check that no orders arrive on taker after the initial one.

On this branch the problem is solved, whenever I restart the taker I do see the trace removing the old connection and receive orders:


```
2021-11-15 13:16:46 TRACE removing outdated connection to taker because unable to send order: Some(Order { id: OrderId(Hyphenated(5a890edf-2719-4250-82de-22d7e3ed8b37)), trading_pair: BtcUsd, position: Short, price: Price(66527.69), min_quantity: Usd(10), max_quantity: Usd(100), leverage: Leverage(2), liquidation_price: Price(44351.793333333333333333333333), creation_timestamp: Timestamp(1636942606), settlement_time_interval_hours: Duration { seconds: 86400, nanoseconds: 0 }, origin: Ours, oracle_event_id: BitMexPriceEventId { timestamp: OffsetDateTime { utc_datetime: PrimitiveDateTime { date: Date { year: 2021, ordinal: 320 }, time: Time { hour: 3, minute: 0, second: 0, nanosecond: 0 } }, offset: UtcOffset { hours: 0, minutes: 0, seconds: 0 } }, digits: 20 } }) taker_id=c0c0ba85-8626-4607-8a97-243f045723a6
2021-11-15 13:16:47 TRACE send to socket message message_name=CurrentOrder
```

**Comment by da-kami** [comment]
It's still scary that we did never see the error before! @thomaseizinger any idea why the error in `handle_broadcast_order` was swallowed?

**Comment by klochowicz** [comment]
> > Did you try if this was the source of our problems?
> 
> Was able to reproduce the problem locally on `master` - steps:
> 
>     1. start maker
> 
>     2. start taker
> 
>     3. check that orders arrive on taker
> 
>     4. keep maker running, restart taker
> 
>     5. check that no orders arrive on taker after the initial one.
> 
> 
> On this branch the problem is solved, whenever I restart the taker I do see the trace removing the old connection and receive orders:
> 
> ```
> 2021-11-15 13:16:46 TRACE removing outdated connection to taker because unable to send order: Some(Order { id: OrderId(Hyphenated(5a890edf-2719-4250-82de-22d7e3ed8b37)), trading_pair: BtcUsd, position: Short, price: Price(66527.69), min_quantity: Usd(10), max_quantity: Usd(100), leverage: Leverage(2), liquidation_price: Price(44351.793333333333333333333333), creation_timestamp: Timestamp(1636942606), settlement_time_interval_hours: Duration { seconds: 86400, nanoseconds: 0 }, origin: Ours, oracle_event_id: BitMexPriceEventId { timestamp: OffsetDateTime { utc_datetime: PrimitiveDateTime { date: Date { year: 2021, ordinal: 320 }, time: Time { hour: 3, minute: 0, second: 0, nanosecond: 0 } }, offset: UtcOffset { hours: 0, minutes: 0, seconds: 0 } }, digits: 20 } }) taker_id=c0c0ba85-8626-4607-8a97-243f045723a6
> 2021-11-15 13:16:47 TRACE send to socket message message_name=CurrentOrder
> ```

shall we try to cover this one with a test? seems like it shouldn't be hard to write one

**Comment by da-kami** [comment]
> > > Did you try if this was the source of our problems?
> > 
> > 
> > Was able to reproduce the problem locally on `master` - steps:
> > ```
> > 1. start maker
> > 
> > 2. start taker
> > 
> > 3. check that orders arrive on taker
> > 
> > 4. keep maker running, restart taker
> > 
> > 5. check that no orders arrive on taker after the initial one.
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > On this branch the problem is solved, whenever I restart the taker I do see the trace removing the old connection and receive orders:
> > ```
> > 2021-11-15 13:16:46 TRACE removing outdated connection to taker because unable to send order: Some(Order { id: OrderId(Hyphenated(5a890edf-2719-4250-82de-22d7e3ed8b37)), trading_pair: BtcUsd, position: Short, price: Price(66527.69), min_quantity: Usd(10), max_quantity: Usd(100), leverage: Leverage(2), liquidation_price: Price(44351.793333333333333333333333), creation_timestamp: Timestamp(1636942606), settlement_time_interval_hours: Duration { seconds: 86400, nanoseconds: 0 }, origin: Ours, oracle_event_id: BitMexPriceEventId { timestamp: OffsetDateTime { utc_datetime: PrimitiveDateTime { date: Date { year: 2021, ordinal: 320 }, time: Time { hour: 3, minute: 0, second: 0, nanosecond: 0 } }, offset: UtcOffset { hours: 0, minutes: 0, seconds: 0 } }, digits: 20 } }) taker_id=c0c0ba85-8626-4607-8a97-243f045723a6
> > 2021-11-15 13:16:47 TRACE send to socket message message_name=CurrentOrder
> > ```
> 
> shall we try to cover this one with a test? seems like it shouldn't be hard to write one

we would need restart behaviour in the test. Not sure how easily we can get that in. Would definitely vote for doing it in a follow up. This problem is blocking me atm 😅

File: daemon/src/maker_inc_connections.rs [hunk 2]
```diff
@@ -157,9 +157,16 @@ impl Actor {
     async fn handle_broadcast_order(&mut self, msg: BroadcastOrder) -> Result<()> {
         let order = msg.0;
         for (taker_id, conn) in self.write_connections.clone() {
-            tracing::trace!(%taker_id, "sending new order for broadcast to connection: {:?}", order);
-            conn.do_send_async(wire::MakerToTaker::CurrentOrder(order.clone()))
-                .await?;
+            if conn
+                .do_send_async(wire::MakerToTaker::CurrentOrder(order.clone()))
+                .await
+                .is_err()
```
Comment (thomaseizinger) [inline daemon/src/maker_inc_connections.rs:163]: I think it would be good to encapsulate this into a function that we use everytime we send something.

**Pull request merged**